apollo-router/src/services/tower_subgraph_service.rs

```diff
@@ -142,3 +164,126 @@ impl tower::Service<graphql::SubgraphRequest> for TowerSubgraphService {
         })
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::convert::Infallible;
+    use std::net::SocketAddr;
+    use std::str::FromStr;
+
+    use axum::Server;
+    use http::header::HOST;
+    use http::Uri;
+    use hyper::service::make_service_fn;
+    use hyper::Body;
+    use tower::{service_fn, ServiceExt};
+
+    use crate::fetch::OperationKind;
+    use crate::{http_compat, Context, Request, Response, SubgraphRequest};
+
+    use super::*;
+
+    // starts a local server emulating a subgraph returning status code 400
+    async fn emulate_subgraph_bad_request(socket_addr: SocketAddr) {
+        async fn handle(_request: http::Request<Body>) -> Result<http::Response<Body>, Infallible> {
+            Ok(http::Response::builder()
+                .header("Content-Type", "application/json")
+                .status(StatusCode::BAD_REQUEST)
+                .body(r#"BAD REQUEST"#.into())
+                .unwrap())
+        }
+
+        let make_svc = make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(handle)) });
+        let server = Server::bind(&socket_addr).serve(make_svc);
+        if let Err(e) = server.await {
+            eprintln!("server error: {}", e);
+        }
+    }
+
+    // starts a local server emulating a subgraph returning bad response format
+    async fn emulate_subgraph_bad_response_format(socket_addr: SocketAddr) {
```

Comment:
do we plan on using both functions in tests only ? or do you wanna provide them to plugin users? not sure